compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -529,6 +529,32 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             obligation.predicate.to_poly_trait_ref(),
             trait_ref,
         )?);
+
+        let lang_items = self.tcx().lang_items();
+        if [lang_items.fn_once_trait(), lang_items.fn_trait(), lang_items.fn_mut_trait()]
+            .contains(&Some(obligation.predicate.def_id()))
+            // Skip `MiscObligation`s for better output.
+            && matches!(obligation.cause.code, ObligationCauseCode::BindingObligation(_, _))
+        {
+            // Do not allow `foo::<fn() -> A>();` for `A: !Sized` (#82633)
+            let fn_sig = obligation.predicate.self_ty().skip_binder().fn_sig(self.tcx());
+            let ty = fn_sig.output().skip_binder();
+            let trait_ref = ty::TraitRef {
+                def_id: lang_items.sized_trait().unwrap(),
+                substs: self.tcx().mk_substs_trait(ty, &[]),
+            };
+            if !matches!(ty.kind(), ty::Projection(_) | ty::Opaque(..)) {
```

Comment:
For reviewer context, the filtering out of projections and opaque types is a hack and shouldn't be necessary long term if we find a better place to inject the Sized obligation.

If I don't keep this in, then we start failing stage 1 builds. This tells me there's a better place to inject this obligation, but I haven't been able to find it.